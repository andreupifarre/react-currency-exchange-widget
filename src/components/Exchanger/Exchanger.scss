.exchanger {
  display: flex;
  flex-direction: column;

  &__bar {
    display: flex;
    height: 0;
    height: 0;
  }

  &__submit-button {
    background-color: #bdd6ee;
    border: 0;
    padding: 18px;
    color: white;
    text-align: center;
    border-radius: 15px;
    font-size: 19px;
    font-family: Roboto, Segoe UI, Tahoma, sans-serif;
    font-weight: 500;
    letter-spacing: 0.5px;
    cursor: pointer;

    &--enabled {
      background-color: #0075eb;
    }

    &:hover {
      opacity: 0.9;
    }

    &:focus {
      outline: 0;
    }
  }
}
