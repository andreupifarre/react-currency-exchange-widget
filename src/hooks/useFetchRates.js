import { useState, useEffect } from 'react';

const useFetchRates = (fromCurrency, interval) => {
  const [rates, setRates] = useState(null);

  useEffect(() => {
    const fetchData = async () => {
      const url = `https://openexchangerates.org/api/latest.json?app_id=81fa12e332b74f45a1de4a276db7c0d1&base=${fromCurrency}`;
      const response = await fetch(url);
      const data = await response.json();
      setRates(data.rates);
    };

    fetchData();
  }, [fromCurrency, interval]);

  return rates;
};

export default useFetchRates;
