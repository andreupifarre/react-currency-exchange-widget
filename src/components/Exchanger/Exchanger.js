import React, { useState, useEffect } from 'react';
import classNames from 'classnames';
import useInterval from '../../hooks/useInterval';
import useBalances from '../../hooks/useBalances';
import useFetchRates from '../../hooks/useFetchRates';
import ExchangerRow from '../ExchangerRow/ExchangerRow';
import ExchangerBar from '../ExchangerBar/ExchangerBar';

import './Exchanger.scss';

export const DEFAULT_FROM_CURRENCY = 'GBP';
export const DEFAULT_TO_CURRENCY = 'EUR';
export const FETCH_INTERVAL = 10000;

const Exchanger = () => {
  const balances = useBalances();
  const [fromCurrency, setFromCurrency] = useState(DEFAULT_FROM_CURRENCY);
  const [toCurrency, setToCurrency] = useState(DEFAULT_TO_CURRENCY);
  const rates = useFetchRates(fromCurrency, useInterval(FETCH_INTERVAL));
  const [fromAmount, setFromAmount] = useState();
  const [toAmount, setToAmount] = useState();
  const [isExceeding, setIsExceeding] = useState(true);
  const [fromInfo, setFromInfo] = useState('');
  const [toInfo, setToInfo] = useState('');
  const [currencies, setCurrencies] = useState(balances.get());

  const selectedFromCurrency = currencies.find((c) => c.code === fromCurrency);
  const selectedToCurrency = currencies.find((c) => c.code === toCurrency);

  const convertFromTo = (currency, amount) =>
    rates ? balances.convertFromTo(amount, rates[currency]) : '';

  const convertToFrom = (currency, amount) =>
    rates ? balances.convertToFrom(amount, rates[currency]) : '';

  const calculate = (amount, currency) => {
    const { isNotAllowedToExchange, fromInfo, toInfo } = balances.calculate(
      amount,
      currency.balance,
      currency.sign,
    );

    setIsExceeding(isNotAllowedToExchange);
    setFromInfo(fromInfo);
    setToInfo(toInfo);
  };

  const handleSwap = () => {
    setFromCurrency(toCurrency);
    setToCurrency(fromCurrency);
    setFromAmount(toAmount);
    setToAmount(fromAmount);
    calculate(toAmount, selectedToCurrency);
  };

  const handleFromCurrencyChange = (currency) => {
    if (currency === toCurrency) {
      setToCurrency(fromCurrency);
      calculate(toAmount, selectedToCurrency);
    }

    setFromCurrency(currency);
  };

  const handleToCurrencyChange = (currency) => {
    if (currency === fromCurrency) {
      setFromCurrency(toCurrency);
      calculate(toAmount, selectedToCurrency);
    }

    setToCurrency(currency);
    setToAmount(convertFromTo(currency, fromAmount));
  };

  const handleFromAmountChange = (amount) => {
    setFromAmount(amount);
    setToAmount(convertFromTo(toCurrency, amount));
    calculate(amount, selectedFromCurrency);
  };

  const handleToAmountChange = (amount) => {
    setToAmount(amount);
    setFromAmount(convertToFrom(toCurrency, amount));
    calculate(amount, selectedFromCurrency);
  };

  const handleExchange = () => {
    const newCurrencies = balances.exchange(
      currencies,
      fromCurrency,
      toCurrency,
      fromAmount,
      toAmount,
    );

    setCurrencies(newCurrencies);
    setFromAmount();
    setToAmount();
    setIsExceeding(true);
  };

  useEffect(() => {
    setToAmount(convertFromTo(toCurrency, fromAmount));
    // eslint-disable-next-line
  }, [rates]);

  return (
    <div className="exchanger">
      <div className="exchanger__row">
        <ExchangerRow
          balance={selectedFromCurrency.balance}
          currency={selectedFromCurrency.code}
          sign={selectedFromCurrency.sign}
          amount={fromAmount}
          info={fromInfo}
          currencies={currencies}
          onAmountChange={handleFromAmountChange}
          onCurrencyChange={handleFromCurrencyChange}
          autoFocus
          alt
        />
      </div>
      <div className="exchanger__bar">
        <ExchangerBar
          onSwap={handleSwap}
          fromSign={selectedFromCurrency.sign}
          toSign={selectedToCurrency.sign}
          rate={rates && rates[toCurrency]}
        />
      </div>
      <div className="exchanger__row">
        <ExchangerRow
          balance={selectedToCurrency.balance}
          currency={selectedToCurrency.code}
          sign={selectedToCurrency.sign}
          amount={toAmount}
          info={toInfo}
          currencies={currencies}
          onAmountChange={handleToAmountChange}
          onCurrencyChange={handleToCurrencyChange}
        />
      </div>
      <button
        className={classNames('exchanger__submit-button', {
          'exchanger__submit-button--enabled': !isExceeding,
        })}
        disabled={isExceeding}
        type="button"
        onClick={handleExchange}
      >
        Exchange
      </button>
    </div>
  );
};

export default Exchanger;
