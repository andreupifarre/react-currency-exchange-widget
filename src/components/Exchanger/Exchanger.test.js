import React from 'react';
import { shallow, render as renderE } from 'enzyme';

import Exchanger from './Exchanger';
import useInterval from '../../hooks/useInterval';
import useFetchRates from '../../hooks/useFetchRates';

jest.mock('../../hooks/useInterval');
jest.mock('../../hooks/useFetchRates');

useInterval.mockReturnValue(0);
useFetchRates.mockReturnValue({});

const render = () => shallow(<Exchanger />);

afterEach(() => {
  jest.clearAllMocks();
});

describe('Exchanger', () => {
  it('should contain a parent element with a class .exchanger', () => {
    const element = render().find('.exchanger');
    expect(element.exists()).toBeTruthy();
  });

  it('should contain a button element with a class .exchanger__submit-button', () => {
    const element = render().find('.exchanger__submit-button');
    expect(element.exists()).toBeTruthy();
    expect(element.name()).toEqual('button');
  });

  it('should contain two ExchangerRow components', () => {
    const wrapper = render();

    const element = wrapper.find('ExchangerRow');
    expect(element.length).toEqual(2);
  });

  it('should not allow to exchange and show excess info when trying to exchange amount higher than balance', () => {
    const rates = {
      GBP: 1,
      EUR: 1.102125,
      USD: 1.293682,
    };
    useFetchRates.mockReturnValue(rates);

    const wrapper = render();

    let row1 = wrapper.find('ExchangerRow').at(0).dive();
    const amount1 = row1.find('.exchanger-row__amount');
    amount1.simulate('change', { target: { value: '12' } });

    row1 = wrapper.find('ExchangerRow').at(0).dive();
    expect(row1.find('.exchanger-row__fee').text()).toEqual('exceeds balance');

    let row2 = wrapper.find('ExchangerRow').at(1).dive();
    expect(row2.find('.exchanger-row__amount').prop('value')).toEqual('13.23');

    expect(wrapper.find('.exchanger__submit-button').prop('disabled'))
      .toBeTruthy;
  });

  it('should correctly exchange currencies when amount is lower than balance', () => {
    const rates = {
      GBP: 1,
      EUR: 1.102125,
      USD: 1.293682,
    };
    useFetchRates.mockReturnValue(rates);

    const wrapper = render();

    let row1 = wrapper.find('ExchangerRow').at(0).dive();
    const amount1 = row1.find('.exchanger-row__amount');
    amount1.simulate('change', { target: { value: '9' } });

    row1 = wrapper.find('ExchangerRow').at(0).dive();
    expect(row1.find('.exchanger-row__fee').text()).toEqual('');

    let row2 = wrapper.find('ExchangerRow').at(1).dive();
    expect(row2.find('.exchanger-row__amount').prop('value')).toEqual('9.92');

    expect(wrapper.find('.exchanger__submit-button').prop('disabled'))
      .toBeFalsy;

    wrapper.find('.exchanger__submit-button').simulate('click');

    expect(
      wrapper
        .find('ExchangerRow')
        .at(0)
        .dive()
        .find('.exchanger-row__balance')
        .text(),
    ).toEqual('Balance: £1.00');

    expect(
      wrapper
        .find('ExchangerRow')
        .at(1)
        .dive()
        .find('.exchanger-row__balance')
        .text(),
    ).toEqual('Balance: €9.92');
  });

  it('should render correctly', () => {
    expect(render().getElements()).toMatchSnapshot();
  });
});
