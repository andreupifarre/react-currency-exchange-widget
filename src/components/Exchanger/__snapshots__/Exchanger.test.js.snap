// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exchanger should render correctly 1`] = `
Array [
  <div
    className="exchanger"
  >
    <div
      className="exchanger__row"
    >
      <ExchangerRow
        alt={true}
        amount=""
        autoFocus={true}
        balance="10"
        currencies={
          Array [
            Object {
              "balance": "10",
              "code": "GBP",
              "sign": "£",
            },
            Object {
              "balance": "0",
              "code": "EUR",
              "sign": "€",
            },
            Object {
              "balance": "0",
              "code": "USD",
              "sign": "$",
            },
          ]
        }
        currency="GBP"
        info=""
        onAmountChange={[Function]}
        onCurrencyChange={[Function]}
        sign="£"
      />
    </div>
    <div
      className="exchanger__bar"
    >
      <ExchangerBar
        fromSign="£"
        onSwap={[Function]}
        rate={1.102125}
        toSign="€"
      />
    </div>
    <div
      className="exchanger__row"
    >
      <ExchangerRow
        alt={false}
        amount=""
        autoFocus={false}
        balance="0"
        currencies={
          Array [
            Object {
              "balance": "10",
              "code": "GBP",
              "sign": "£",
            },
            Object {
              "balance": "0",
              "code": "EUR",
              "sign": "€",
            },
            Object {
              "balance": "0",
              "code": "USD",
              "sign": "$",
            },
          ]
        }
        currency="EUR"
        info=""
        onAmountChange={[Function]}
        onCurrencyChange={[Function]}
        sign="€"
      />
    </div>
    <button
      className="exchanger__submit-button"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      Exchange
    </button>
  </div>,
]
`;
